### DISCUSSION TITLE
File search results in rofi

### DISCUSSION THREAD

**Discussion by CapelaGames** [davatorium/rofi#1894] [Q&A] [answered]
Hi,
I'm trying to get a file search working, and feeding those results to rofi.

I am currently using a python script however I'm new to python, (C developer jumping into the deep end). I would love some help on what I am doing wrong.

I tried looking up documentation and doing some searches. 

I'm using `rofi  -show find -modi find:~/.local/share/rofi/finder.py` to run the script below, it however, doesn't show any results in  rofi. I tried printing rofi_input like this `print(rofi_input)` , which shows results in rofi, but doesn't allow me to open any file.

```
#!/usr/bin/env python3

import subprocess

if __name__ == '__main__':
    # Use fd to get a list of files and directories, and split them into a list
    files = subprocess.run(['fd'], encoding='utf-8', capture_output=True).stdout.splitlines()

    # Set the title for the Rofi window
    title = 'Find Files'

    # Construct the rofi command using dmenu mode, fuzzy search (-i), and the provided title
    rofi_command = ['rofi', '-dmenu', '-i', '-p', title]

    # Join the list of files into a single string, separated by newline characters
    rofi_input = '\n'.join(file for file in files)

    # Run Rofi with the constructed command and the file list as input
    cp = subprocess.run(rofi_command, input=rofi_input,encoding='utf-8', capture_output=True)

    
    # If an item was selected in Rofi
    if cp.returncode == 0:
        cp.check_returncode()

        selected_file = cp.stdout.strip()

        # Use xdg-open to open the selected file with the default associated application
        xdg_command = ['xdg-open', selected_file]
        subprocess.run(xdg_command, encoding='utf-8', capture_output=True)

```
Thank you for any help!

**Comment by DaveDavenport** [comment]
Why not use the build in file-browser? `rofi -show filebrowser -modes filebrowser`

**Reply by DaveDavenport** [reply]
Btw.. you are mixing up rofi in dmenu emulation mode and rofi's script mode. 
This will never work. 

If you want to make a script, I suggest you read the `rofi-script` manpage, if you want to make a dmenu script, I recommend the many online tutorials and the `rofi-dmenu` manual page.

**Reply by CapelaGames** [reply]
> Why not use the build in file-browser?

I want to do a fuzzy search on all my files, which fd allows. As far as I understand, file-browser only allows searching within the directory your in.

> Btw.. you are mixing up rofi in dmenu emulation mode and rofi's script mode.

Thanks so much, that's a huge help, I will look into this. 

I will leave this unanswered for now and reply with my solution later.

**Reply by DaveDavenport** [reply]
In git there is `recursivebrowser` it searches (nicely async) in all files in a directory.

**Reply by CapelaGames** [reply]
Oh nice, I will check that out. Thank you for taking time to help me.

**Comment by CapelaGames** [comment, marked as answer]
I'm not sure if there is an easier way to do this but this is how I got it working.
(replace username with your username, I couldn't seem to get ~ working)
Use this to run the following script  .`/finder.py`

```
#!/usr/bin/env python3

import subprocess
import shlex

if __name__ == '__main__':

    # arg1 = sys.argv[1]
    # Use fd to get a list of files and directories, and split them into a list
    files = subprocess.run(shlex.split("fd . /home/username"), encoding='utf-8', capture_output=True).stdout.splitlines()

    # Set the title for the Rofi window
    title = 'Find Files'

    # Construct the rofi command using dmenu 
    rofi_command = ['rofi', '-dmenu', '-i', '-p', title]

    # Join the list of files into a single string, separated by newline characters
    rofi_input = '\n'.join(file for file in files)

    # Run Rofi with the constructed command and the file list as input
    cp = subprocess.run(rofi_command, input=rofi_input,encoding='utf-8', capture_output=True)  

    # If an item was selected in Rofi
    if cp.returncode == 0:
        cp.check_returncode()

        selected_file = cp.stdout.strip()

        # Use xdg-open to open the selected file with the default associated application
        xdg_command = ['xdg-open', selected_file]
        subprocess.run(xdg_command, encoding='utf-8', capture_output=True)
```
